Assets/ShooterProject/Scripts/Weapons/Weapon.cs

```diff
@@ -0,0 +1,152 @@
+using ShooterProject.Scripts.General;
+using System.Collections;
+using UnityEngine;
+using UnityEngine.InputSystem;
+using UnityEngine.XR.Interaction.Toolkit;
+
+namespace ShooterProject.Scripts.Weapons
+{
+
+	public class Weapon : MonoBehaviour
+	{
+		#region Fields
+
+		[SerializeField]
+		private WeaponParts _weaponParts;
+
+		[SerializeField]
+		private WeaponParams _weaponParams;
+
+		[SerializeField]
+		private WeaponShootingEffects _weaponShootingEffects;
+
+		private bool _isSelected = false;
+		private bool _coolDownOver = true;
+		private Coroutine _workingShootingCoroutine;
+		private GameObjectsPool _impactsPool;
+
+		#endregion
+
+		#region Properties
+		public bool IsSelected => _isSelected;
+		private bool _canShoot => _isSelected;
+		#endregion
+
+		#region LifeCycle Methods
+		private void Start()
+		{
+			_impactsPool = new GameObjectsPool(_weaponShootingEffects.MaxImpacts, false, false, _weaponShootingEffects.ImpactPrefab, null);
+		}
+		private void OnEnable()
+		{
+			AddEventsListeners();
+		}
+
+		private void OnDisable()
+		{
+			RemoveEventsListeners();
+		}
+
+		#endregion
+
+		#region Private Methods
+
+		private IEnumerator ShootingCoroutine()
+		{
+			while (_canShoot)
+			{
+
+				if (!_coolDownOver)
+				{
+					yield return new WaitForEndOfFrame();
+					continue;
+				}
+
+				SingleShot();
+
+				PlaySound(_weaponShootingEffects.ShootignSound);
+
+				StartCoroutine(ShootingCoolDownCoroutine());
+
+				if (!_weaponParams.CanFireBursts)
+					yield break;
+			}
+		}
+
+		private void SingleShot()
+		{
+			RaycastHit hitInfo;
+			Vector3 weaponForward = _weaponParts.BulletSpawner.forward;
+
+			if (Physics.Raycast(_weaponParts.BulletSpawner.position, weaponForward, out hitInfo, _weaponParams.ShootingDistance))
```

Comment:
Можно сделать out параметр в одной строке 
```suggestion
			if (Physics.Raycast(_weaponParts.BulletSpawner.position, weaponForward, out var hitInfo, _weaponParams.ShootingDistance))
```